filters/include/pcl/filters/lambda_filter_indices.h

```diff
@@ -0,0 +1,166 @@
+                                           pcl::index_t>,
+                     bool>;
+
+// can't use this for SFINAE since Derived isn't properly defined
+// but this can be used after the class definition to test it
+/*
+template <class Base, class Derived>
+constexpr auto IsValidLambdaFilter = std::enable_if_t<
+    std::is_base_of<Base, Derived>::value &&
+        pcl::is_invocable_v<std::declval<Derived>().get_lambda, void> &&
+        std::is_same<PointFilterLambda<std::declval<Derived>().get_lambda()>,
+bool>, bool>;
+*/
+
+/**
+ * \brief LambdaFilterIndices filters point clouds and indices based on a
+ * function pointer passed to filter command \ingroup filters
+ */
+template <typename PointT, typename Derived>
+struct LambdaFilterIndicesImpl : public FilterIndices<PointT> {
+private:
+  using Base = FilterIndices<PointT>;
+  using PCLBase = pcl::PCLBase<PointT>;
+
+protected:
+  using PointCloud = typename FilterIndices<PointT>::PointCloud;
+  using PointCloudPtr = typename PointCloud::Ptr;
+  using PointCloudConstPtr = typename PointCloud::ConstPtr;
+  using FieldList = typename pcl::traits::fieldList<PointT>::type;
+
+  using Base::extract_removed_indices_;
+  using Base::removed_indices_;
+  using PCLBase::indices_;
+  using PCLBase::input_;
+
+protected:
+  // to prevent instantiation of impl class
+  LambdaFilterIndicesImpl(bool extract_removed_indices)
+  : FilterIndices<PointT>(extract_removed_indices)
+  {}
+  /** \brief Filtered results are indexed by an indices array.
+   * \param[out] indices The resultant indices.
+   */
+  void
+  applyFilter(std::vector<int>& indices) override
+  {
+    indices.clear();
+    indices.reserve(input_->points.size());
+    if (extract_removed_indices_) {
+      removed_indices_->clear();
+      removed_indices_->reserve(input_->points.size());
+    }
+
+    const auto& lambda = static_cast<Derived*>(this)->get_lambda();
+
+    for (const auto index : *indices_) {
+      // lambda returns true for points that shoud be selected
+      if (!(this->negative_) == lambda(input_->points[index], index)) {
+        indices.push_back(index);
+      }
+      else {
+        if (extract_removed_indices_) {
+          removed_indices_->push_back(index);
+        }
+      }
+    }
+  }
+};
+} // namespace detail
+
+template <typename PointT, typename Functor>
+struct LambdaFilterIndices
+: public detail::LambdaFilterIndicesImpl<PointT, LambdaFilterIndices<PointT, Functor>> {
+private:
+  using Self = LambdaFilterIndices<PointT, Functor>;
+  using Base = detail::LambdaFilterIndicesImpl<PointT, Self>;
+
+public:
+  using FunctorT = Functor;
+  // using in type would complicate signature
+  static_assert(std::is_same<detail::PointFilterLambda<PointT, FunctorT>, bool>::value,
+                "Functor needs to be able to satisfy the callable constraint");
+
+protected:
+  using Base::filter_name_;
+  // need to hold a value because lambdas can only be copy or move constructed
+  const FunctorT lambda_;
```

Comment:
Did you want to take a reference here? If yes, wouldn't `std::reference_wrapper` solve your issue?

Reply:
No, that wouldn't resolve the issue. That'd bring in more issues regarding lifetime.

Reply:
<details><summary>Source Code</summary>

```cpp
#include <pcl/filters/lambda_filter_indices.h>
#include <pcl/point_cloud.h>
#include <pcl/point_types.h>
#include <pcl/types.h>

#include <benchmark/benchmark.h>

#include <functional>
#include <random>

struct PointCloudFixture : public benchmark::Fixture {
  void
  SetUp(const ::benchmark::State& state)
  {
    cloud = pcl::make_shared<pcl::PointCloud<pcl::PointXYZ>>();
    cloud->resize(state.range(0));
    for (auto& pt : *cloud) {
      pt = pcl::PointXYZ{0., 0., 0.};
    }
  }
  pcl::shared_ptr<pcl::PointCloud<pcl::PointXYZ>> cloud;
};

BENCHMARK_DEFINE_F(PointCloudFixture, FunctionTest)(benchmark::State& st)
{
  const auto lambda = [](const pcl::PointXYZ& pt, pcl::index_t) {
    return (pt.getArray3fMap() < 10).all() && (pt.getArray3fMap() > -10).all();
  };

  const std::function<bool(const pcl::PointXYZ&, pcl::index_t)> functor = lambda;
  pcl::LambdaFilterIndices<pcl::PointXYZ, decltype(functor)> filter{functor};
  filter.setInputCloud(cloud);

  auto out_cloud = pcl::make_shared<pcl::PointCloud<pcl::PointXYZ>>();
  for (auto _ : st) {
    filter.filter(*out_cloud);
  }
}

BENCHMARK_DEFINE_F(PointCloudFixture, LambdaTest)(benchmark::State& st)
{
  const auto lambda = [](const pcl::PointXYZ& pt, pcl::index_t) {
    return (pt.getArray3fMap() < 10).all() && (pt.getArray3fMap() > -10).all();
  };

  pcl::LambdaFilterIndices<pcl::PointXYZ, decltype(lambda)> filter{lambda};
  filter.setInputCloud(cloud);

  auto out_cloud = pcl::make_shared<pcl::PointCloud<pcl::PointXYZ>>();
  for (auto _ : st) {
    filter.filter(*out_cloud);
  }
}

BENCHMARK_REGISTER_F(PointCloudFixture, FunctionTest)->Range(8, 8<<10);
BENCHMARK_REGISTER_F(PointCloudFixture, LambdaTest)->Range(8, 8<<10);

BENCHMARK_MAIN();
```
</details>

```
Run on (8 X 1700 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 256 KiB (x4)
  L3 Unified 8192 KiB (x1)
Load Average: 2.50, 5.10, 10.73
------------------------------------------------------------------------------
Benchmark                                    Time             CPU   Iterations
------------------------------------------------------------------------------
PointCloudFixture/FunctionTest/8           154 ns          153 ns      4664894
PointCloudFixture/FunctionTest/64          401 ns          400 ns      1579348
PointCloudFixture/FunctionTest/512        2094 ns         2091 ns       326307
PointCloudFixture/FunctionTest/4096      16322 ns        16293 ns        40278
PointCloudFixture/FunctionTest/8192      33009 ns        32952 ns        21205
PointCloudFixture/LambdaTest/8             110 ns          109 ns      5791490
PointCloudFixture/LambdaTest/64            326 ns          326 ns      2034351
PointCloudFixture/LambdaTest/512          1586 ns         1584 ns       390011
PointCloudFixture/LambdaTest/4096        12528 ns        12488 ns        51640
PointCloudFixture/LambdaTest/8192        24642 ns        24597 ns        27225
```

Reply:
Using `std::function` gives a 0.75x speedup 😄

Reply:
Well, the benchmark settles this discussion. Since this is a hot loop, the use of lambda really shines through.